File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.java
import java.util.*;
class Solution {
    
    private Set<String> opSet;
    
    private String[] operands = {"+", "-", "*"};
    private boolean[] visited = {false, false, false};
    
    private void makeOperandSet(String t) {
        if (t.length() == 3) {
            opSet.add(t);
            return;
        }
        
        for (int i = 0; i < 3; i++) {
            if (!visited[i]) {
                visited[i] = true;
                makeOperandSet(t + operands[i]);
                visited[i] = false;
            }  
        } 
    }
    
    	
    private String calc(String num1, String op, String num2) {
        long n1 = Long.parseLong(num1);
        long n2 = Long.parseLong(num2);
 
        if (op.equals("+"))
            return n1 + n2 + "";
        else if (op.equals("-"))
            return n1 - n2 + "";
        return n1 * n2 + "";
    }
    
    public long solution(String expression) {
        long answer = 0;
        
        opSet = new HashSet<>();
        String[] split = expression.split("(?<=[-+*/])|(?=[-+*/])");
        
        makeOperandSet("");
        
        long max = Long.MIN_VALUE;
        long min = Long.MAX_VALUE;
        
        for (String op : opSet) {
            List<String> list = new ArrayList<>(List.of(split));
            for (char o : op.toCharArray()) {
                for (int i = 0; i < list.size(); i++) {
                    if (list.get(i).equals(String.valueOf(o))) {
                        list.set(i-1, calc(list.get(i-1), list.get(i), list.get(i+1)));
                        list.remove(i);
                        list.remove(i);
                        i--;
                    }   
                }
            }
            answer = Math.max(answer, Math.abs(Long.parseLong(list.get(0))));
        }
        
        return answer;
    }
}

File: 프로그래머스/2/67257. ［카카오 인턴］ 수식 최대화/［카카오 인턴］ 수식 최대화.py
import re
from itertools import permutations

def solution(expression):
    tokens = re.split("([-+*])", expression)
    operators = ['+', '-', '*']
    answer = 0

    for p in permutations(operators):
        temp_tokens = tokens[:]
        for op in p:
            stack = []
            i = 0
            while i < len(temp_tokens):
                if temp_tokens[i] == op:
                    left = stack.pop()
                    right = temp_tokens[i+1]
                    res = str(eval(left + op + right))
                    stack.append(res)
                    i += 2 
                else:
                    stack.append(temp_tokens[i])
                    i += 1
            temp_tokens = stack
        answer = max(answer, abs(int(temp_tokens[0])))

    return answer